src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.Reducer.cs

```diff
@@ -0,0 +1,349 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    using static SyntaxFactory;
+
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider
+    {
+        private sealed class Reducer
+        {
+            // Track whether we have made progress by merging nodes.
+            // If this remains false, we will not offer the fix.
+            private bool _isNonTrivial;
+
+            private Reducer() { }
+
+            // The reducer, pairs each node with each other and check if the combination is valid.
+            public static AnalyzedNode Reduce(AnalyzedNode node, out bool isNonTrivial)
+            {
+                var reducer = new Reducer();
+                var result = reducer.Reduce(node);
+                isNonTrivial = reducer._isNonTrivial;
+                return result;
+            }
+
+            private AnalyzedNode Intersection(AnalyzedNode left, AnalyzedNode right)
```

Comment:
what does 'intersection' mean in this context?

Reply:
think I should revert the old naming here. it will be VisitConjunction, which is what it really is.